crates/rpc/rpc-eth-types/src/cache/mod.rs

```diff
@@ -253,6 +254,30 @@ impl EthStateCache {
         Ok(block.zip(receipts))
     }
 
+    /// Fetches receipts and optionally fetches the block if available in the cache.
+    /// If the block is not in the cache and is not near the tip, only returns the receipts.
+    pub async fn get_receipts_and_maybe_block(
+        &self,
+        block_num_hash: &BlockNumHash,
+        best_number: u64,
+    ) -> ProviderResult<Option<(Arc<Vec<Receipt>>, Option<SealedBlock>)>> {
+        // Check if the block is near the tip (within the last 4 blocks)
+        let cached_range = best_number.saturating_sub(4)..=best_number;
+
+        if cached_range.contains(&block_num_hash.number) {
+            // Fetch both receipts and block if within the tip range
+            let block = self.get_sealed_block(block_num_hash.hash);
+            let receipts = self.get_receipts(block_num_hash.hash);
+
+            let (block, receipts) = futures::try_join!(block, receipts)?;
+
+            Ok(receipts.map(|r| (r, block)))
+        } else {
+            // Fetch only receipts if the block is not within the tip range
+            Ok(self.get_receipts(block_num_hash.hash).await?.map(|r| (r, None)))
+        }
```

Comment:
I’ve rolled back the changes to move the tip-check logic back to filter.rs and adjusted the cache function to return receipts even if the block isn’t cached. This ensures the cache function now uses map instead of zip, so it’s more flexible in returning just the receipts if the block is unavailable.

I think that this might feel like a subtle change, and I'm wondering if this refactor fully meets the original goals of the issue. Could you confirm whether this approach aligns with what you had in mind, or if you feel further adjustments (or none at all) are needed?